services/wireguard/connection/connection.go

```diff
@@ -165,7 +165,7 @@ func (c *Connection) start(ctx context.Context, start startConn, options connect
 	log.Info().Msgf("Adding connection peer %s", config.Provider.Endpoint.String())
 
 	log.Info().Msg("Waiting for initial handshake")
-	if err := c.handshakeWaiter.Wait(conn.PeerStats, c.opts.HandshakeTimeout, c.done); err != nil {
+	if err := c.handshakeWaiter.Wait(ctx, conn.PeerStats, c.opts.HandshakeTimeout, c.done); err != nil {
```

Comment:
This will now return an error if context is canceled, but will it get handled by:
```
	defer func() {
		if err != nil {
			c.Stop()
		}
	}()
```

Which is above this code? I think that it wont because `err` here is a local variable to the `if` statement. Not sure about the consequences of not calling `c.Stop` on error though, maybe @soffokl  can comment on that.

Reply:
in the logs, I see that the Stop function gets called when the connection is cancelled, so it seems to work.

Reply:
The stop you see probably called outside. This shadowed `err` in the if will not trigger the condition in the defer. 
this might be fixed by: 
```suggestion
	if err = c.handshakeWaiter.Wait(ctx, conn.PeerStats, c.opts.HandshakeTimeout, c.done); err != nil {
```

Reply:
And also change all the `err:=` to `err=` in previous lines too (except the first one), right?

Reply:
The first one can be changed too, you already have it declared in the function.